### ISSUE TITLE
Quarkus test IllegalAccessError

### ISSUE THREAD

**Issue by andrzejszywala** [quarkusio/quarkus#18553]
## Describe the bug
In quarkus test I've mocked CDI bean which is in separate jar file. When I run test in which this mocked bean is used I've error

`
java.lang.IllegalAccessError: class com.example.ExampleRepositoryMock_Bean tried to access field com.example.BaseRepository.someValue (com.example.ExampleRepositoryMock_Bean is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @6bc269bf; com.example.BaseRepository is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @1976e6ca)
        at com.example.ExampleServiceTest.test(ExampleServiceTest.java:20)
`

In quarkus 1.13.7 it works fine. 

### Expected behavior
Test should pass

### Actual behavior
Exception 

`
java.lang.IllegalAccessError: class com.example.ExampleRepositoryMock_Bean tried to access field com.example.BaseRepository.someValue (com.example.ExampleRepositoryMock_Bean is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @6bc269bf; com.example.BaseRepository is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @1976e6ca)
        at com.example.ExampleServiceTest.test(ExampleServiceTest.java:20)
`

## To Reproduce

Link to a small reproducer (preferably a Maven project if the issue is not Gradle-specific).

Or attach an archive containing the reproducer to the issue.

1.  extract attached file [illegal_access_bug.zip](https://github.com/quarkusio/quarkus/files/6789837/illegal_access_bug.zip)
2.  open folder `commons`
3.  run `mvn install'
4.  open folder `app`
5.  mvnw test

### Configuration
```properties
# Add your application.properties here, if applicable.

```

### Screenshots
(If applicable, add screenshots to help explain your problem.)

### Output of `uname -a` or `ver`
icrosoft Windows [Version 10.0.18363.1556]

### Output of `java -version`
openjdk version "11.0.9.1" 2020-11-04 LTS
OpenJDK Runtime Environment Corretto-11.0.9.12.1 (build 11.0.9.1+12-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.9.12.1 (build 11.0.9.1+12-LTS, mixed mode)

### Quarkus version or git rev
2.0.1.Final
### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: C:\Users\aszywala\.m2\wrapper\dists\apache-maven-3.8.1-bin\2l5mhf2pq2clrde7f7qp1rdt5m\apache-maven-3.8.1
Java version: 11.0.9.1, vendor: Amazon.com Inc., runtime: c:\jdk\jdk11.0.9_12
Default locale: pl_PL, platform encoding: Cp1250
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
## Additional context
(Add any other context about the problem here.)

**Comment by famod** [comment]
This is a split package case since both the `BaseRepository` in module `common` _and_ `ExampleRepositoryMock` in module `app` share the same package `com.example`.
I personally would not consider this best practise.

Anyway, /cc-ing you @mkouba @manovotn because @geoand added the ArC label.

**Comment by famod** [comment]
PS: The problem goes away if `ExampleRepositoryMock` is placed in a different package.

**Comment by manovotn** [comment]
@famod is right that split package is a bad practice. I'd suggest just changing one of the packages.

Furthermore, this will only manifest is some Quarkus modes - test being one, dev the other. In prod it should work because of flat class loader.

It is similar (in nature) to what we discussed few days ago here - https://github.com/quarkusio/quarkus/issues/18510
The above issue also raised a question of detection of this setup and whether it's worth the effort. I looked at the code and didn't see a way to be able to tell, from within Quarkus, that two classes with same package are from different JARs. I might be missing something though, so if anyone has a clue, do let me know :-)
